tests/feedback.test.ts

```diff
@@ -0,0 +1,184 @@
+});
+
+afterAll(async () => {
+  await DatabaseConnection.clear();
+  await DatabaseConnection.close();
+});
+
+describe('feedback submission', () => {
+  test('properly persists on successful submission', async () => {
+    const conn = await DatabaseConnection.get();
+    const [user] = UserFactory.create(1);
+    const [feedback] = FeedbackFactory.create(1);
+
+    await new PortalState().createUsers([user]).write(conn);
+
+    const feedbackController = ControllerFactory.feedback(conn);
+    await feedbackController.submitFeedback({ feedback }, user);
+    const submittedFeedbackResponse = await feedbackController.getFeedback(user);
+
+    expect(submittedFeedbackResponse.feedback).toHaveLength(1);
+    expect(submittedFeedbackResponse.feedback[0]).toStrictEqual({
+      ...submittedFeedbackResponse.feedback[0],
+      user: user.getPublicProfile(),
+      status: FeedbackStatus.SUBMITTED,
+      ...feedback,
+    });
+  });
+
+  test('is invalidated when submission description is too short', async () => {
+    const [feedback] = FeedbackFactory.with({ description: 'A short description' });
+
+    const errors = await validate(plainToClass(Feedback, feedback));
+
+    expect(errors).toBeDefined();
+    expect(errors).toHaveLength(1);
+    expect(errors[0].property).toEqual('description');
+    expect(errors[0].constraints.minLength).toBeDefined();
+  });
+
+  test('has has proper activity scope and type', async () => {
+    const conn = await DatabaseConnection.get();
+    const [user] = UserFactory.create(1);
+    const [feedback] = FeedbackFactory.create(1);
+
+    await new PortalState().createUsers([user]).write(conn);
+
+    await ControllerFactory.feedback(conn).submitFeedback({ feedback }, user);
+    const activityResponse = await ControllerFactory.user(conn).getCurrentUserActivityStream(user);
+    const feedbackSubmissionActivity = activityResponse.activity[1];
+
+    expect(feedbackSubmissionActivity.scope).toEqual(ActivityScope.PRIVATE);
+    expect(feedbackSubmissionActivity.type).toEqual(ActivityType.SUBMIT_FEEDBACK);
+  });
+
+  test('admins can view feedback from any user', async () => {
+    const conn = await DatabaseConnection.get();
+    const [user1, user2] = UserFactory.create(2);
+    const [admin] = UserFactory.with({ accessType: UserAccessType.ADMIN });
+    const [feedback1, feedback2] = FeedbackFactory.create(2);
+
+    await new PortalState().createUsers([user1, user2, admin]).write(conn);
+
+    const feedbackController = ControllerFactory.feedback(conn);
+    const submittedFeedback1Response = await feedbackController.submitFeedback({ feedback: feedback1 }, user1);
+    const submittedFeedback2Response = await feedbackController.submitFeedback({ feedback: feedback2 }, user2);
+    const allSubmittedFeedbackResponse = await feedbackController.getFeedback(admin);
+
+    expect(allSubmittedFeedbackResponse.feedback).toHaveLength(2);
+    expect(allSubmittedFeedbackResponse.feedback).toEqual(
```

Comment:
Can we do a `toStrictEqual([...])` here (and in the other array assertions) or do we need to use the `arrayContaining`?

Reply:
Nope don't think so, since the database tables have indices on uuid, they are pulled in sorted order by uuid which is unpredictable (since feedback factory doesn't produce models)